web/src/topic_list.ts

```diff
@@ -181,12 +192,32 @@ export function is_complete_for_stream_id(stream_id: number): boolean {
             return true;
         }
 
-        // Now, we can just compare the first cached message to the first
-        // message ID in the stream; if it's older, we're good, otherwise,
-        // we might be missing the oldest topics in this stream in our
-        // cache.
         const first_cached_message = all_messages_data.first_including_muted();
-        return first_cached_message!.id <= sub.first_message_id;
+        if (sub.first_message_id < first_cached_message!.id) {
+            // Missing the oldest topics in this stream in our cache.
+            return false;
+        }
+
+        // At this stage, number of topics displayed in the topic list
+        // widget is at max `topic_list_data.max_topics` and
+        // sub.first_message_id >= first_cached_message!.id.
+        //
+        // There's a possibility of a few topics missing for messages
+        // which were sent when the user wasn't subscribed.
+        // Fetch stream history to confirm if all topics are already
+        // displayed otherwise rebuild with updated data.
+        stream_topic_history_util.get_server_history(stream_id, () => {
+            const history = stream_topic_history.find_or_create(stream_id);
+            if (history.topics.size > num_topics_displayed) {
+                update_widget_for_stream(stream_id);
+            }
+        });
+        // We return `false` which leads to visible 'show all topics',
+        // even if all the topics are already displayed.
+        // This is helpful if the API call fails, users will have
+        // the option to make another request. Otherwise there's
+        // a possibility of missing 'show all topics' & not all topics displayed.
+        return false;
```

Comment:
Just highlighting the UX decision here, let me know if it's worth discussing in #design.

The reason of returning `false` is explained in the code as comments.

Note if the API call doesn't fail and `history.topics.size === num_topics_displayed`, then 'show all topics' will stay there even though we're displaying all topics. The solution is to `update_widget_for_stream(stream_id)` unconditionally, but I think it's fine to show 'show all topics' & avoid doing unnecessary `.build()` in other cases.

Reply:
I think this UX is good.